### PR TITLE
[Feature] Support minicpmv v2.6

### PR DESCRIPTION
**Pull request by mickqian**
## Motivation

Addressing #2461 

## Modifications

1. Add new model `MiniCPMV` and corresponding processor `MiniCPMVImageProcessor`
2. Register new chat and conversation templates

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contribution_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contribution_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by mickqian** [comment]
## Questions
1. From my understanding of current implementation of sglang, it does not support two kind of attentions at the same time, thus the attention module in `Resampler` can't be replaced with `RadixAttention`. Is that correct?

**Comment by zhaochenyang20** [comment]
> ## Questions
> 1. From my understanding on current implementation of sglang, it does not support two kind of attentions at the same time, thus the attention module in `Resampler` and `Idefics2VisionAttention` can't be replaced with `RadixAttention`. Is that correct?
> 2. Current implementation of `Idefics2VisionAttention` in this PR relies on `vllm.attention.MultiheadAttention`, which is introduced in version `vllm==0.6.5`. If that is not acceptable, I will rewrite the attention module

@Ying1123 @merrymercy

**Comment by zhaochenyang20** [comment]
@mickqian Looks nice! Are there anything left to do?

**Comment by mickqian** [comment]
> @mickqian Looks nice! Are there anything left to do?看起来不错！还有什么要做的吗？

1. The number of frames processed from the video is hard-coded for the moment, which should be calculated. I added a TODO on it.
2. A CI Test failed because of this:
```
[2025-01-10 03:24:51 TP0] Multimodal prompt is too long after expanding multimodal tokens. After expanding len(req.origin_input_ids_unpadded)=29 => 308 >= 294.
```
from an assertion of checking the input_ids' length. The token limit (from my understanding) is calculated from the GPU: 
```python
 if len(req.origin_input_ids) >= self.max_req_input_len:
        logger.error(
            "Multimodal prompt is too long after expanding multimodal tokens. "
            f"After expanding {len(req.origin_input_ids_unpadded)=} => {len(req.origin_input_ids)} >= {self.max_req_input_len}. "
                )
```
3. Some mp4 file failed the decode process in `decode_video_base64`

**Comment by zhaochenyang20** [comment]
@hnyls2002 Could you help to review. Thanks!

File: python/sglang/srt/models/llava.py [hunk 1]
```diff
@@ -403,7 +403,7 @@ def forward(
                     pt += 1
 
             return self.language_model(
-                input_ids, positions, forward_batch, input_embeds=input_embeds
```
Comment (yizhang2077) [inline python/sglang/srt/models/llava.py]: Maybe we can keep the name "input_embeds" since it can keep the same as "input_ids" ?

File: python/sglang/srt/models/minicpmv.py [hunk 2]
```diff
@@ -0,0 +1,1335 @@
...
+            self.embed_dim,
+            self.embed_dim,
+            bias=True,
+            quant_config=quant_config,
+            prefix=f"{prefix}.out_proj",
+        )
+        self.tp_size = get_tensor_model_parallel_world_size()
+        self.num_heads_per_partition = divide(self.num_heads, self.tp_size)
+        # TODO: does forward_batch supports two attentions at the same time?
+        # self.attn = RadixAttention(num_heads=self.num_heads_per_partition,
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmv.py]: Currently it does not support to use RadixAttention without casual mask, maybe you can try to use `context_attention_fwd` as qwen2_vl does or wrap it with a easier module, and then you do not need to write MHA

**Comment by zhaochenyang20** [comment]
Hey @mickqian , I asked Yi Zhang to help. Gonna get back soon

**Comment by yizhang2077** [comment]
> > ## Questions
> > 
> > 1. From my understanding on current implementation of sglang, it does not support two kind of attentions at the same time, thus the attention module in `Resampler` and `Idefics2VisionAttention` can't be replaced with `RadixAttention`. Is that correct?
> > 2. Current implementation of `Idefics2VisionAttention` in this PR relies on `vllm.attention.MultiheadAttention`, which is introduced in version `vllm==0.6.5`. If that is not acceptable, I will rewrite the attention module

1. I think you can use  `RadixAttention` at the same time(save_kv_cache = False). But `RadixAttention` do not support run without casual mask. In qwen2_vl, it use `context_attention_fwd` instead. I think maybe it is better to wrap it with MultiheadAttention module so that other multimodal llm can reuse it.

File: python/sglang/srt/managers/image_processor.py [hunk 3]
```diff
@@ -229,6 +232,126 @@ async def process_images_async(
...
+
+        image_hashes, image_sizes = [], []
+        raw_images = []
+        IMAGE_TOKEN = "(<image>./</image>)"
+
+        def encode_video(video_path):
+            if not os.path.exists(video_path):
+                return []
+            # TODO
+            MAX_NUM_FRAMES = 3  # if cuda OOM set a smaller number
```
Comment (yizhang2077) [inline python/sglang/srt/managers/image_processor.py]: it means minicpmv only support 3 frames for a video?

Comment (mickqian) [reply python/sglang/srt/managers/image_processor.py]: No, I'm doing so only because I don't know about the policy of calculating the desired number of frames. Should we just allow it to read the entire video?

Comment (yizhang2077) [reply python/sglang/srt/managers/image_processor.py]:
```
 if len(req.origin_input_ids) >= self.max_req_input_len:
        logger.error(
            "Multimodal prompt is too long after expanding multimodal tokens. "
            f"After expanding {len(req.origin_input_ids_unpadded)=} => {len(req.origin_input_ids)} >= {self.max_req_input_len}. "
                )
```
If read the entire video, is this assertion easy to fail? If so, I think we can adjust max num frames according to `self.max_req_input_len`.

**Comment by zhaochenyang20** [comment]
@mickqian Thanks. We will update the API key as soon as possible. Before that, could you fix the conflicts of `qwen2_vl.py`.

**Comment by yizhang2077** [comment]
LGTM cc @merrymercy

**Review by yizhang2077 — APPROVED** [review]

File: python/sglang/srt/managers/schedule_batch.py [hunk 4]
```diff
@@ -149,6 +147,16 @@ class ImageInputs:
...
     mrope_position_delta: Optional[torch.Tensor] = None
 
+    # MiniCPMV related
+    # All the images in the batch should share the same special image
+    # bound token ids.
+    im_start_id: Optional[torch.Tensor] = None
+    im_end_id: Optional[torch.Tensor] = None
+    slice_start_id: Optional[torch.Tensor] = None
+    slice_end_id: Optional[torch.Tensor] = None
+
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py:157]: remove this empty line

File: python/sglang/srt/models/minicpmv.py [hunk 5]
```diff
@@ -0,0 +1,1238 @@
...
+import torch.types
+from PIL import Image
+from torch import nn
+from torch.nn.init import trunc_normal_
+from transformers import PretrainedConfig
+from vllm.distributed import divide, get_tensor_model_parallel_world_size
+from vllm.model_executor.layers.resampler import get_2d_sincos_pos_embed
+from vllm.model_executor.layers.sampler import SamplerOutput, get_sampler
+from vllm.model_executor.models.module_mapping import MultiModelKeys
+from vllm.model_executor.sampling_metadata import SamplingMetadata
```
Comment (merrymercy) [inline python/sglang/srt/models/minicpmv.py:46]: Do not import anything from vllm.
- `SamplingMetadata`, `SamplerOutput` are not used.
- Use sglang.srt.distributed
- copy over small utility functions.

File: python/sglang/srt/utils.py [hunk 6]
```diff
@@ -451,6 +451,8 @@ def load_image(image_file: Union[str, bytes]):
     else:
         raise ValueError(f"Invalid image: {image}")
 
+    # if image_size is None:
```
Comment (merrymercy) [inline python/sglang/srt/utils.py:454]: remove this

**Review by merrymercy — COMMENTED** [review]
1. Can you add some unit tests like this to compare the logtis against HF implementation? https://github.com/sgl-project/sglang/pull/2365/files#diff-6e52783df34170e0b3d9aadbd7c338d9c16f0303c18846afc1d298c32d4a4eb2R1
2. Can you help update the docs here for VLM? https://github.com/sgl-project/sglang/blob/83452dbb4a19c6a2461e972eb2b64a2df9a466b8/docs/references/supported_models.md?plain=1#L50

**Comment by zhaochenyang20** [comment]
@mickqian Also, remove Chinese comments in the PR. Thanks so much. Could we try not import any vllm dependency, rather rewrite it ourselves.

**Comment by zhaochenyang20** [comment]
Under the models files, we prefer not to import anything from vllm. We will remove them all later.

**Comment by mickqian** [comment]
Trying to address these in #2977
